### ISSUE TITLE
QLoRA on CPU - Example ERROR: "undefined symbol"

### ISSUE THREAD

**Issue by 42elenz** [intel/intel-extension-for-transformers#1287]
Hello! I am trying to use the QLoRa for CPU.
Unfortunately I ran into this Error now:

> /python3.11/site-packages/intel_extension_for_transformers/libqbits.so: undefined symbol: _ZNR5torch7Library4_defEON3c106eitherINS1_12OperatorNameENS1_14FunctionSchemaEEEONS_11CppFunctionE

Can someone help me with that? 
here is my code. I exchanged decapoda with baffo32/decapoda since the model mentioned is not available anymore.

`import torch
from intel_extension_for_transformers.transformers.modeling import AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained(
    'baffo32/decapoda-research-llama-7B-hf',
    torch_dtype=torch.bfloat16,
    load_in_4bit=True,
    use_llm_runtime=False
)
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training, TaskType
model = prepare_model_for_kbit_training(
    model, use_gradient_checkpointing=True
)
model.gradient_checkpointing_enable()
peft_config = LoraConfig(
    r=8,
    task_type=TaskType.CAUSAL_LM,
)
model = get_peft_model(model, peft_config)`

This is my env:
name: qlora_intel_tutorial
channels:
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=5.1=1_gnu
  - asttokens=2.4.1=pyhd8ed1ab_0
  - bzip2=1.0.8=h7b6447c_0
  - ca-certificates=2024.2.2=hbcca054_0
  - comm=0.2.1=pyhd8ed1ab_0
  - debugpy=1.6.7=py311h6a678d5_0
  - decorator=5.1.1=pyhd8ed1ab_0
  - exceptiongroup=1.2.0=pyhd8ed1ab_2
  - executing=2.0.1=pyhd8ed1ab_0
  - importlib-metadata=7.0.1=pyha770c72_0
  - importlib_metadata=7.0.1=hd8ed1ab_0
  - ipykernel=6.29.2=pyhd33586a_0
  - ipython=8.21.0=pyh707e725_0
  - jedi=0.19.1=pyhd8ed1ab_0
  - jupyter_client=8.6.0=pyhd8ed1ab_0
  - jupyter_core=5.7.1=py311h38be061_0
  - ld_impl_linux-64=2.38=h1181459_1
  - libffi=3.4.4=h6a678d5_0
  - libgcc-ng=11.2.0=h1234567_1
  - libgomp=11.2.0=h1234567_1
  - libsodium=1.0.18=h36c2ea0_1
  - libstdcxx-ng=11.2.0=h1234567_1
  - libuuid=1.41.5=h5eee18b_0
  - matplotlib-inline=0.1.6=pyhd8ed1ab_0
  - ncurses=6.4=h6a678d5_0
  - nest-asyncio=1.6.0=pyhd8ed1ab_0
  - openssl=3.0.13=h7f8727e_0
  - packaging=23.2=pyhd8ed1ab_0
  - parso=0.8.3=pyhd8ed1ab_0
  - pexpect=4.9.0=pyhd8ed1ab_0
  - pickleshare=0.7.5=py_1003
  - pip=23.3.1=py311h06a4308_0
  - platformdirs=4.2.0=pyhd8ed1ab_0
  - prompt-toolkit=3.0.42=pyha770c72_0
  - ptyprocess=0.7.0=pyhd3deb0d_0
  - pure_eval=0.2.2=pyhd8ed1ab_0
  - pygments=2.17.2=pyhd8ed1ab_0
  - python=3.11.7=h955ad1f_0
  - python-dateutil=2.8.2=pyhd8ed1ab_0
  - python_abi=3.11=2_cp311
  - pyzmq=25.1.2=py311h6a678d5_0
  - readline=8.2=h5eee18b_0
  - setuptools=68.2.2=py311h06a4308_0
  - six=1.16.0=pyh6c4a22f_0
  - sqlite=3.41.2=h5eee18b_0
  - stack_data=0.6.2=pyhd8ed1ab_0
  - tk=8.6.12=h1ccaba5_0
  - tornado=6.3.3=py311h5eee18b_0
  - traitlets=5.14.1=pyhd8ed1ab_0
  - typing_extensions=4.9.0=pyha770c72_0
  - wcwidth=0.2.13=pyhd8ed1ab_0
  - wheel=0.41.2=py311h06a4308_0
  - xz=5.4.5=h5eee18b_0
  - zeromq=4.3.5=h6a678d5_0
  - zipp=3.17.0=pyhd8ed1ab_0
  - zlib=1.2.13=h5eee18b_0
  - pip:
    - accelerate==0.27.2
    - aiohttp==3.9.3
    - aiosignal==1.3.1
    - attrs==23.2.0
    - bitsandbytes==0.42.0
    - certifi==2024.2.2
    - charset-normalizer==3.3.2
    - contextlib2==21.6.0
    - contourpy==1.2.0
    - cycler==0.12.1
    - datasets==2.17.1
    - deprecated==1.2.14
    - dill==0.3.8
    - docstring-parser==0.15
    - filelock==3.13.1
    - fonttools==4.49.0
    - frozenlist==1.4.1
    - fsspec==2023.10.0
    - huggingface-hub==0.20.3
    - idna==3.6
    - intel-extension-for-pytorch==2.2.0
    - intel-extension-for-transformers==1.3.1
    - ipywidgets==8.1.2
    - jinja2==3.1.3
    - joblib==1.3.2
    - jupyterlab-widgets==3.0.10
    - kiwisolver==1.4.5
    - markdown-it-py==3.0.0
    - markupsafe==2.1.5
    - matplotlib==3.8.3
    - mdurl==0.1.2
    - mpmath==1.3.0
    - multidict==6.0.5
    - multiprocess==0.70.16
    - networkx==3.2.1
    - neural-compressor==2.4.1
    - neural-speed==0.2
    - numpy==1.26.4
    - nvidia-cublas-cu12==12.1.3.1
    - nvidia-cuda-cupti-cu12==12.1.105
    - nvidia-cuda-nvrtc-cu12==12.1.105
    - nvidia-cuda-runtime-cu12==12.1.105
    - nvidia-cudnn-cu12==8.9.2.26
    - nvidia-cufft-cu12==11.0.2.54
    - nvidia-curand-cu12==10.3.2.106
    - nvidia-cusolver-cu12==11.4.5.107
    - nvidia-cusparse-cu12==12.1.0.106
    - nvidia-nccl-cu12==2.19.3
    - nvidia-nvjitlink-cu12==12.3.101
    - nvidia-nvtx-cu12==12.1.105
    - opencv-python-headless==4.9.0.80
    - pandas==2.2.0
    - peft==0.8.2
    - pillow==10.2.0
    - prettytable==3.9.0
    - psutil==5.9.8
    - py-cpuinfo==9.0.0
    - pyarrow==15.0.0
    - pyarrow-hotfix==0.6
    - pycocotools==2.0.7
    - pyparsing==3.1.1
    - pytz==2024.1
    - pyyaml==6.0.1
    - regex==2023.12.25
    - requests==2.31.0
    - rich==13.7.0
    - safetensors==0.4.2
    - schema==0.7.5
    - scikit-learn==1.4.1.post1
    - scipy==1.12.0
    - sentencepiece==0.1.99
    - shtab==1.6.5
    - sympy==1.12
    - threadpoolctl==3.3.0
    - tokenizers==0.15.2
    - torch==2.2.0
    - tqdm==4.66.2
    - transformers==4.37.2
    - triton==2.2.0
    - trl==0.7.11
    - tyro==0.7.3
    - tzdata==2024.1
    - urllib3==2.2.1
    - widgetsnbextension==4.0.10
    - wrapt==1.16.0
    - xxhash==3.4.1
    - yarl==1.9.4

**Comment by zhewang1-intc** [comment]
hi, if you install ITREX by pip, then libqbits.so compiled based on 2.1.0+cpu.
pls reinstall correct version torch or build ITREX from source.

**Comment by 42elenz** [comment]
Thank you very much for your fast answer. I now have this Error, which is indicating that my cpus are missing the AMX-BF16 support. Is there a way to still use the library? 

> Qbits: device ISA must support AMX-BF16 when compute_type==bf16

Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         46 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  32
  On-line CPU(s) list:   0-31
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
    CPU family:          15
    Model:               6
    Thread(s) per core:  1
    Core(s) per socket:  1
    Socket(s):           32
    Stepping:            1
    BogoMIPS:            5985.93
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp
                          lm constant_tsc nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_d
                         eadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp i
                         brs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd av
                         x512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities
Virtualization features: 
  Hypervisor vendor:     KVM
  Virtualization type:   full
Caches (sum of all):     
  L1d:                   1 MiB (32 instances)
  L1i:                   1 MiB (32 instances)
  L2:                    128 MiB (32 instances)
  L3:                    512 MiB (32 instances)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-31
Vulnerabilities:         
  Gather data sampling:  Not affected
  Itlb multihit:         KVM: Mitigation: VMX unsupported
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Mmio stale data:       Unknown: No mitigations
  Retbleed:              Not affected
  Spec rstack overflow:  Not affected
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl and seccomp
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
  Srbds:                 Not affected
  Tsx async abort:       Mitigation; TSX disabled

**Comment by zhewang1-intc** [comment]
yes, you can use fp32 compute_type which using AVX_512F to compute GEMM. But this may need to modify some source code. @XinyuYe-Intel can provide some help.

**Comment by 42elenz** [comment]
Thank you! I was able to train it with by just jusing torch.float32. 
Last problem I have:
Is there a way to use multiple CPUs? At the moment it is just using one CPU which is really slowly. 

`import torch
from intel_extension_for_transformers.transformers.modeling import AutoModelForCausalLM
from datasets import load_dataset
from peft import get_peft_model,  LoraConfig, TaskType, prepare_model_for_kbit_training
from transformers import (
    AutoTokenizer,
    TrainingArguments
)

from trl import SFTTrainer

model_name = "mistralai/Mistral-7B-v0.1"
#Tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name, add_eos_token=True, use_fast=True)
tokenizer.pad_token = tokenizer.unk_token
tokenizer.pad_token_id =  tokenizer.unk_token_id
tokenizer.padding_side = 'left'

model = AutoModelForCausalLM.from_pretrained(
          model_name, load_in_4bit=True, use_llm_runtime=False, torch_dtype=torch.float32, low_cpu_mem_usage=False
)
model = prepare_model_for_kbit_training(model, use_gradient_checkpointing = True)
#Configure the pad token in the model
model.config.pad_token_id = tokenizer.pad_token_id

model.gradient_checkpointing_enable()
peft_config = LoraConfig(
    r=8,
    task_type=TaskType.CAUSAL_LM,
)

model = get_peft_model(model, peft_config)

[...]

training_arguments = TrainingArguments(
        output_dir="./results_sft_cpu/",
        #evaluation_strategy="steps",
        #do_eval=True,
        per_device_train_batch_size=2,
        gradient_accumulation_steps=2,
        per_device_eval_batch_size=4,
        log_level="debug",
        save_steps=20,
        logging_steps=10,
        learning_rate=1e-4,
        #eval_steps=10,
        max_steps=200,
        warmup_steps=20,
        lr_scheduler_type="linear",
)

trainer = SFTTrainer(
        model=model,
        train_dataset=dataset_train_sft,
        #\val_dataset=dataset_test_sft,
        dataset_text_field="text",
        max_seq_length=256,
        tokenizer=tokenizer,
        args=training_arguments,
)

trainer.train()
`

For example with the accelerate library?

**Comment by zhewang1-intc** [comment]
yes, you can use Intel® oneCCL library to leverage multi-CPUs compute capability to accelerate model training.
